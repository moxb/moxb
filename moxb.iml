<?xml version="1.0" encoding="UTF-8"?>
<module type="WEB_MODULE" version="4">
  <component name="NewModuleRootManager" inherit-compiler-output="true">
    <exclude-output />
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/packages/antd/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/meteor/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/moxb/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/react-html/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/examples/src" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/packages/moxb/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/meteor/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/react-html/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/antd/dist" />
      <excludeFolder url="file://$MODULE_DIR$/admin/node-installation" />
      <excludeFolder url="file://$MODULE_DIR$/admin/bin-tools" />
      <excludeFolder url="file://$MODULE_DIR$/examples/dist" />
      <excludeFolder url="file://$MODULE_DIR$/examples/.cache" />
      <excludeFolder url="file://$MODULE_DIR$/coverage" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>