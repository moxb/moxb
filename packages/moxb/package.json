{
  "name": "@moxb/moxb",
  "version": "0.2.0-beta.59",
  "description": "MobX data binding",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "main": "dist",
  "typings": "dist/",
  "files": [
    "dist/**/*"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/moxb/moxb.git"
  },
  "author": "moxb (moxb.org)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/moxb/moxb/issues"
  },
  "homepage": "https://github.com/moxb/moxb#readme",
  "keywords": [
    "mobx",
    "UI"
  ],
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "transform": {
      "^.+\\.(ts|tsx)$": "ts-jest"
    },
    "testEnvironment": "node",
    "testMatch": [
      "**/src/**/__tests__/*.(ts|tsx)"
    ],
    "modulePaths": [
      "<rootDir>/peer-dependencies/node_modules"
    ]
  },
  "peerDependencies": {
    "mobx": "^5.15.6",
    "tslib": "^2.0.1"
  },
  "devDependencies": {
    "@types/atob": "2.1.2",
    "@types/jest": "26.0.13",
    "@types/lodash.flatten": "4.4.6",
    "@types/urijs": "1.19.11",
    "mobx": "5.13.1",
    "tslib": "1.10.0"
  },
  "dependencies": {
    "btoa-lite": "1.0.0",
    "history": "5.0.0",
    "lodash.flatten": "4.4.0",
    "moment": "2.29.1",
    "urijs": "1.19.2"
  }
}
