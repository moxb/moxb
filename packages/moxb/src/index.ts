export * from './impl';
export * from './types';
export * from './routing';
