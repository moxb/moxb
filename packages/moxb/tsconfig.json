{
    "compilerOptions": {
        "module": "commonjs",
        "target": "ES5",
        "jsx": "react",
        "lib": ["dom","es6","es2015","es2017", "dom.iterable","scripthost"],
        "inlineSourceMap": true,
        "importHelpers":true,
        "experimentalDecorators": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strict": true,
        "outDir": "./dist",
        "declaration": true,
        "skipLibCheck": true,
        "strictNullChecks": true,
        "removeComments": false,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "allowUnreachableCode": false,
        "baseUrl": "./src",
        "paths": {
            "*": [
                "*",
                "../../../node_modules/*"
            ]
        }
    },
    "exclude": [
        "./node_modules",
        "./dist"
    ]
}
