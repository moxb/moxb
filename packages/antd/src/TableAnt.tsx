import { observer } from 'mobx-react';
import * as React from 'react';
import { Table as MoxTable, t } from '@moxb/moxb';
import { Alert, Table } from 'antd';
import { ColumnProps, TableProps } from 'antd/lib/table/interface';
import { TextSearchAnt } from './TextAnt';

export interface ColumnAntProps<T> extends ColumnProps<T> {
    dataIndex: string;
    column: string;
}

export interface TableAntProps<T> extends TableProps<any> {
    table: MoxTable<T>;
    hideHeader?: boolean;
    /**
     * Setup the column
     * @param column
     */
    setupColumn?(column: ColumnAntProps<T>): void;
}

function toCell(x: any, _record: any) {
    if (React.isValidElement(x)) {
        return x;
    }
    return x + '';
}
@observer
export class TableAnt<T> extends React.Component<TableAntProps<T>> {
    render() {
        const { table, ...tableProps } = this.props;
        const columns = table.columns.map((column: any) => ({
            column: column.column,
            title: column.label,
            dataIndex: column.column,
            key: column.tableColumn,
            sorter: column.sortable,
            render: toCell,
        }));
        if (this.props.setupColumn) {
            columns.forEach((column: any) => this.props.setupColumn!(column));
        }
        const dataSource = table.data.map((data: any, idx: number) => ({ key: idx + '', ...(data as any) }));
        return (
            <>
                {table.errors!.length > 0 && (
                    <Alert message={t('Table.Error.title', 'Error')} description={table.errors} type="error" />
                )}
                {table.search && (
                    <TextSearchAnt
                        required
                        style={{ marginBottom: '1.5em' }}
                        enterButton={t('Table.Search.title', 'Search')}
                        operation={table.search.searchField}
                        onSearch={() => table.search!.searchAction.fire()}
                    />
                )}
                <Table
                    columns={columns}
                    dataSource={dataSource}
                    loading={!table.ready}
                    pagination={
                        table.pagination
                            ? {
                                  total: table.pagination.totalAmount,
                                  showSizeChanger: true,
                                  showQuickJumper: true,
                                  pageSizeOptions: table.pagination.pageSizes.map(p => '' + p),
                              }
                            : undefined
                    }
                    onChange={(pagination, _filters, sorter) => {
                        if (table.pagination) {
                            if (pagination.pageSize) {
                                table.pagination.setPageSize(pagination.pageSize);
                            }
                            if (pagination.current != null) {
                                table.pagination.setActivePage(pagination.current);
                            }
                        }
                        if (sorter && sorter.columnKey) {
                            table.sort.setSort(
                                sorter.columnKey,
                                sorter.order === 'ascend' ? 'ascending' : 'descending'
                            );
                        }
                        // todo filter
                    }}
                    {...tableProps}
                />
            </>
        );
    }
}
