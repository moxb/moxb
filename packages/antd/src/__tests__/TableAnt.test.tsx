import { TableImpl, TableColumnImpl, TableSearchImpl, TablePaginationImpl } from '@moxb/moxb';
import { shallow } from 'enzyme';
import * as React from 'react';
import { TableAnt, ColumnAntProps } from '../TableAnt';

describe('TableAnt', function() {
    const data = [
        { _id: '1', email: 'john@doe.com', fullName: 'John Doe', createdAt: '2018-01-01' },
        { _id: '2', email: 'johanna@yahoo.com', fullName: 'Johanna Doe', createdAt: '2018-05-01' },
        { _id: '3', email: 'jake@gmail.com', fullName: 'Jake Doe', createdAt: '2018-10-01' },
        { _id: '4', email: 'max@mustermann.com', fullName: 'Max Mustermann', createdAt: '2017-13-07' },
    ];
    function renderColumn(column: ColumnAntProps<any>) {
        column.render = (columnData: any) => columnData.toString();
    }
    function newTableOperation(opt?: any) {
        return new TableImpl(
            Object.assign(
                {
                    id: 'table',
                    data: (tab: any) => tab.sort.sortData(data),
                    columns: (bind: any) => [
                        new TableColumnImpl(
                            {
                                id: 'email',
                                label: 'E-Mail',
                                preferredSortDirection: 'ascending',
                            },
                            bind
                        ),
                        new TableColumnImpl(
                            {
                                id: 'fullName',
                                label: 'Full Name',
                                preferredSortDirection: 'ascending',
                            },
                            bind
                        ),
                        new TableColumnImpl(
                            {
                                id: 'createdAt',
                                label: 'Joined',
                                preferredSortDirection: 'descending',
                            },
                            bind
                        ),
                    ],
                },
                ...opt
            )
        );
    }
    it('should render a table with content by default', function() {
        const operation = newTableOperation();
        expect(shallow(<TableAnt table={operation} />)).toMatchSnapshot();
    });

    it('should show an alert, if an error in the table happened', function() {
        const operation = newTableOperation();
        operation.setError('New error');
        const wrapper = shallow(<TableAnt table={operation} />);
        expect(
            wrapper
                .render()
                .first()
                .find('.ant-alert-message').length
        ).toBe(1);
    });

    it('should render a table with setupColumn set', function() {
        const operation = newTableOperation();
        expect(shallow(<TableAnt table={operation} setupColumn={column => renderColumn(column)} />)).toMatchSnapshot();
    });

    it('should render a table with search field and pagination', function() {
        const operation = newTableOperation({
            search: new TableSearchImpl(),
            pagination: new TablePaginationImpl({
                totalAmount: () => 4,
            }),
        });
        expect(shallow(<TableAnt table={operation} />)).toMatchSnapshot();
    });
});
