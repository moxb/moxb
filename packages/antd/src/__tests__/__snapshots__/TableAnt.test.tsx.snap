// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableAnt should render a table with content by default 1`] = `
<Fragment>
  <Table
    columns={
      Array [
        Object {
          "column": "email",
          "dataIndex": "email",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "email",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "E-Mail",
          "width": undefined,
        },
        Object {
          "column": "fullName",
          "dataIndex": "fullName",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "fullName",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "Full Name",
          "width": undefined,
        },
        Object {
          "column": "createdAt",
          "dataIndex": "createdAt",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "createdAt",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "Joined",
          "width": undefined,
        },
      ]
    }
    data-testid="table"
    dataSource={
      Array [
        Object {
          "_id": "1",
          "createdAt": "2018-01-01",
          "email": "john@doe.com",
          "fullName": "John Doe",
          "key": "0",
        },
        Object {
          "_id": "2",
          "createdAt": "2018-05-01",
          "email": "johanna@yahoo.com",
          "fullName": "Johanna Doe",
          "key": "1",
        },
        Object {
          "_id": "3",
          "createdAt": "2018-10-01",
          "email": "jake@gmail.com",
          "fullName": "Jake Doe",
          "key": "2",
        },
        Object {
          "_id": "4",
          "createdAt": "2017-13-07",
          "email": "max@mustermann.com",
          "fullName": "Max Mustermann",
          "key": "3",
        },
      ]
    }
    loading={false}
    onChange={[Function]}
    rowKey="key"
    scroll={
      Object {
        "x": 1500,
      }
    }
  />
</Fragment>
`;

exports[`TableAnt should render a table with search field and pagination 1`] = `
<Fragment>
  <Table
    columns={
      Array [
        Object {
          "column": "email",
          "dataIndex": "email",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "email",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "E-Mail",
          "width": undefined,
        },
        Object {
          "column": "fullName",
          "dataIndex": "fullName",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "fullName",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "Full Name",
          "width": undefined,
        },
        Object {
          "column": "createdAt",
          "dataIndex": "createdAt",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "createdAt",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "Joined",
          "width": undefined,
        },
      ]
    }
    data-testid="table"
    dataSource={
      Array [
        Object {
          "_id": "1",
          "createdAt": "2018-01-01",
          "email": "john@doe.com",
          "fullName": "John Doe",
          "key": "0",
        },
        Object {
          "_id": "2",
          "createdAt": "2018-05-01",
          "email": "johanna@yahoo.com",
          "fullName": "Johanna Doe",
          "key": "1",
        },
        Object {
          "_id": "3",
          "createdAt": "2018-10-01",
          "email": "jake@gmail.com",
          "fullName": "Jake Doe",
          "key": "2",
        },
        Object {
          "_id": "4",
          "createdAt": "2017-13-07",
          "email": "max@mustermann.com",
          "fullName": "Max Mustermann",
          "key": "3",
        },
      ]
    }
    loading={false}
    onChange={[Function]}
    pagination={
      Object {
        "current": 1,
        "pageSizeOptions": Array [
          "10",
          "25",
          "50",
          "100",
        ],
        "showQuickJumper": true,
        "showSizeChanger": true,
        "total": 4,
      }
    }
    rowKey="key"
    scroll={
      Object {
        "x": 1500,
      }
    }
  />
</Fragment>
`;

exports[`TableAnt should render a table with setupColumn set 1`] = `
<Fragment>
  <Table
    columns={
      Array [
        Object {
          "column": "email",
          "dataIndex": "email",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "email",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "E-Mail",
          "width": undefined,
        },
        Object {
          "column": "fullName",
          "dataIndex": "fullName",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "fullName",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "Full Name",
          "width": undefined,
        },
        Object {
          "column": "createdAt",
          "dataIndex": "createdAt",
          "defaultSortOrder": "descend",
          "fixed": undefined,
          "key": "createdAt",
          "render": [Function],
          "sortOrder": false,
          "sorter": true,
          "title": "Joined",
          "width": undefined,
        },
      ]
    }
    data-testid="table"
    dataSource={
      Array [
        Object {
          "_id": "1",
          "createdAt": "2018-01-01",
          "email": "john@doe.com",
          "fullName": "John Doe",
          "key": "0",
        },
        Object {
          "_id": "2",
          "createdAt": "2018-05-01",
          "email": "johanna@yahoo.com",
          "fullName": "Johanna Doe",
          "key": "1",
        },
        Object {
          "_id": "3",
          "createdAt": "2018-10-01",
          "email": "jake@gmail.com",
          "fullName": "Jake Doe",
          "key": "2",
        },
        Object {
          "_id": "4",
          "createdAt": "2017-13-07",
          "email": "max@mustermann.com",
          "fullName": "Max Mustermann",
          "key": "3",
        },
      ]
    }
    loading={false}
    onChange={[Function]}
    rowKey="key"
    scroll={
      Object {
        "x": 1500,
      }
    }
    setupColumn={[Function]}
  />
</Fragment>
`;
